File: Programa1/Programa1.java
package Programa1;

import java.util.Random;

public class Programa1 {
    public static void main(String[] args) {
        
        if (args.length == 0) {
            System.out.println("Por favor, proporciona el número de palabras como argumento.");
            return;
        }

        int n = 0;
        try {
            n = Integer.parseInt(args[0]);
        } catch (NumberFormatException e) {
            System.out.println("El argumento debe ser un número entero.");
            return;
        }

        char[] cadenota = new char[n * 4]; // 3 letras + 1 espacio por palabra
        String letras = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        String objetivo = "IPN";
        Random rand = new Random();

        long inicio = System.currentTimeMillis();

        // Generar palabras y almacenar en cadenota
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < 3; j++) {
                cadenota[i * 4 + j] = letras.charAt(rand.nextInt(letras.length()));
            }
            cadenota[i * 4 + 3] = ' '; // separador
        }

        int ocurrencias = 0;

        // Buscar "IPN" en cadenota
        for (int i = 0; i < n * 4; i += 4) { // saltamos palabra por palabra
            if (cadenota[i] == 'I' && cadenota[i + 1] == 'P' && cadenota[i + 2] == 'N') {
                ocurrencias++;
                System.out.println("¡Coincidencia encontrada en la posición " + i + "!");
            }
        }

        long fin = System.currentTimeMillis();
        double tiempoTotal = (fin - inicio) / 1000.0;

        System.out.println("\nResumen:");
        System.out.println("Total de ocurrencias de 'IPN': " + ocurrencias);
        System.out.println("Tiempo total transcurrido: " + tiempoTotal + " segundos.");
    }
}

File: Programa1/programa1.py
import random
import sys
import time

def main():
    # Verifica que se haya proporcionado el número de cadenas como argumento
    if len(sys.argv) < 2:
        print("Por favor, proporciona el número de cadenas como argumento.")
        return

    try:
        num_cadenas = int(sys.argv[1])
    except ValueError:
        print("El argumento debe ser un número entero.")
        return

    letras = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    objetivo = "IPN"
    ocurrencias = 0
    tiempos = []
    intentos_ocurrencia = []

    inicio = time.time()

    for i in range(num_cadenas):
        # Genera las primeras 3 letras
        primera_parte = ''.join(random.choice(letras) for _ in range(3))
        # Genera las siguientes 3 letras
        segunda_parte = ''.join(random.choice(letras) for _ in range(3))
        cadena = primera_parte + " " + segunda_parte

        if primera_parte == objetivo:
            ocurrencias += 1
            tiempo_ocurrencia = time.time() - inicio
            tiempos.append(tiempo_ocurrencia)
            intentos_ocurrencia.append(i + 1)
            print(f"¡Coincidencia encontrada: {cadena} en el intento {i + 1}. Tiempo desde inicio: {tiempo_ocurrencia:.3f} segundos.")

    fin = time.time()
    tiempo_total = fin - inicio

    print("\nResumen:")
    print(f"Total de ocurrencias de '{objetivo}': {ocurrencias}")
    if ocurrencias > 0:
        for k in range(ocurrencias):
            print(f"Ocurrencia #{k + 1} en el intento {intentos_ocurrencia[k]}, tiempo: {tiempos[k]:.3f} segundos.")
    else:
        print(f"No se encontró la coincidencia '{objetivo}'.")
    
    print(f"Tiempo total transcurrido: {tiempo_total:.3f} segundos.")

if __name__ == "__main__":
    main()
